### PR TITLE
[DI] Improve error message for non-autowirable scalar argument

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #26838
| License       | MIT
| Doc PR        | -

### CODE REVIEW

File: src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php [hunk 1]
```diff
@@ -233,7 +233,13 @@ private function autowireMethod(\ReflectionFunctionAbstract $reflectionMethod, a
                     if ($parameter->isOptional()) {
                         continue;
                     }
-                    throw new AutowiringFailedException($this->currentId, sprintf('Cannot autowire service "%s": argument "$%s" of method "%s()" must have a type-hint or be given a value explicitly.', $this->currentId, $parameter->name, $class !== $this->currentId ? $class.'::'.$method : $method));
+                    if ($type = ProxyHelper::getTypeHint($reflectionMethod, $parameter, false)) {
+                        $message = 'Cannot autowire service "%s": argument "%s $%s" of method "%s()" must be given a value explicitly.';
+                    } else {
+                        $message = 'Cannot autowire service "%s": argument "%s$%s" of method "%s()" must have a type-hint or be given a value explicitly.';
```
Comment (Tobion) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: just adding a type hint does not neccessarily help like `string`. So maybe it should say something like "must have a known type-hint".

**Comment by nicolas-grekas** [comment]
message updated

File: src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php [hunk 2]
```diff
@@ -233,7 +233,10 @@ private function autowireMethod(\ReflectionFunctionAbstract $reflectionMethod, a
                     if ($parameter->isOptional()) {
                         continue;
                     }
-                    throw new AutowiringFailedException($this->currentId, sprintf('Cannot autowire service "%s": argument "$%s" of method "%s()" must have a type-hint or be given a value explicitly.', $this->currentId, $parameter->name, $class !== $this->currentId ? $class.'::'.$method : $method));
+                    $type = ProxyHelper::getTypeHint($reflectionMethod, $parameter, false);
+                    $type = $type ? sprintf('is type-hinted "%s"', $type) : 'has not type-hint';
```
Comment (Tobion) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: typo: has no type-hint

Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: fixed

**Review by Tobion — APPROVED** [review]

**Review by stof — APPROVED** [review]

**Pull request merged**